pilot/model/controller_view.go

```diff
@@ -0,0 +1,580 @@
+		cv.reconcileService(k, addSvc, EventAdd)
+	}
+	for k, updSvc := range updateSet {
+		cv.reconcileService(k, updSvc, EventUpdate)
+	}
+}
+
+// Reconciles service instances in the ControllerView with the list of expectedInstances
+func (cv *ControllerView) reconcileServiceInstances(expectedInstances []*ServiceInstance) {
+	// No need to synchronize cv.services given this is the only writer thread
+	// All modifications to cv ought to be done after locking cv.mu
+	actualKeyInstMap := make(map[string]*ServiceInstance, len(cv.serviceInstances))
+	for k := range cv.serviceInstances {
+		actualKeyInstMap[k] = cv.serviceInstances[k]
+	}
+
+	expectedKeyInstMap := make(map[string]*ServiceInstance, len(expectedInstances))
+	for _, inst := range expectedInstances {
+		instKey := "[" + inst.Service.Hostname + "][" + getNormalizedIP(inst.Endpoint.Address) + "][" +
+			getNormalizedPort(inst.Endpoint.Port)
+		expectedKeyInstMap[instKey] = inst
+	}
+	updateSet := map[string]*ServiceInstance{}
+	for k, expInst := range expectedKeyInstMap {
+		actInst, found := actualKeyInstMap[k]
+		if !found {
+			continue // Needs to be added to ControllerView
+		}
+		if !reflect.DeepEqual(*expInst, *actInst) {
+			updateSet[k] = expInst
+		}
+		// Remaining would be ones that need adding
+		delete(expectedKeyInstMap, k)
+		// Remaining would be ones that need deleting
+		delete(actualKeyInstMap, k)
+	}
+	cv.mu.Lock()
+	defer cv.mu.Unlock()
+	for k, delInst := range actualKeyInstMap {
+		cv.reconcileServiceInstance(k, delInst, EventDelete)
+	}
+	for k, addInst := range expectedKeyInstMap {
+		cv.reconcileServiceInstance(k, addInst, EventAdd)
+	}
+	for k, updInst := range updateSet {
+		cv.reconcileServiceInstance(k, updInst, EventUpdate)
+	}
+}
+
+// reconcileService is expected to be called only from inside reconcileServices(). The caller is expected
+// to lock the resource view before calling this method()
+func (cv *ControllerView) reconcileService(k string, s *Service, e Event) {
+	old, found := cv.services[k]
+	if found {
+		cv.serviceLabels.deleteLabel(k, labelServiceName, old.Hostname)
+		if old.ExternalName != "" {
+			cv.serviceLabels.deleteLabel(k, labelServiceExternalName, old.ExternalName)
+		}
+		if old.Address != "" {
+			cv.serviceLabels.deleteLabel(k, labelServiceVIP, getNormalizedIP(old.Address))
+		}
+		delete(cv.services, k)
+	}
+	if e != EventDelete {
+		// Treat as upsert
+		cv.serviceLabels.addLabel(k, labelServiceName, s.Hostname)
+		if s.ExternalName != "" {
+			cv.serviceLabels.addLabel(k, labelServiceExternalName, s.ExternalName)
+		}
+		if s.Address != "" {
+			cv.serviceLabels.addLabel(k, labelServiceVIP, getNormalizedIP(s.Address))
+		}
+		cv.services[k] = s
+	}
+}
+
+// reconcileServiceInstance is expected to be called only from inside reconcileServiceInstances(). The caller is expected
+// to lock the resource view before calling this method()
+func (cv *ControllerView) reconcileServiceInstance(k string, i *ServiceInstance, e Event) {
```

Comment:
btw, service instance events are broken. there's no bug open, but it doesn't work reliably on k8s.